### ISSUE TITLE
[Bug] preAllocatedSize could be negative in corner cases

### ISSUE THREAD

**Issue by advancedxy** [apache/incubator-uniffle#426]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

Notice `preAllocatedSize` could be negative in prod server logs

### Affects Version(s)

master

### Uniffle Server Log Output

```logtalk
[INFO] 2022-12-15 10:11:50,921 Grpc-*** ShuffleBufferManager flushIfNecessary - Start to flush with usedMemory[88*******], preAllocatedSize[-4969120], inFlushSize[0]
```

### Additional context

I did some investigation, there's some race conditions to update the `preAllocatedSize` field.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by xianjingfeng** [comment]
Duplicate issue #229

**Comment by jerqi** [comment]
closed by #428
